### ISSUE TITLE
Small object detection

### ISSUE THREAD

**Issue by rui-rich** [ultralytics/ultralytics#20657]
I would like to ask you, is there any way to improve the detection rate of small objects (less than 20*20 pixels)? How to improve the recall of small objects?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @rui-rich, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will also assist you here soon 😊

**Comment by glenn-jocher** [comment]
Hi @rui-rich,

For improving small object detection (< 20x20 pixels), consider these techniques:

1. Image tiling: Split larger images into smaller segments during training and inference to maintain resolution for small objects
2. Increase input resolution: Use `--img 1280` or higher to capture more detail
3. Use mosaic and copy-paste augmentations during training
4. Try models with higher resolution feature maps like YOLO11-l or YOLO11-x
5. Adjust anchor scales if using anchor-based models
6. Fine-tune IoU and confidence thresholds specifically for small objects

The combination of these approaches typically yields significant improvements in small object recall. Let me know if you'd like more details on implementing any of these methods.

**Comment by arun-gautham** [comment]
I have a followup question , what if the small objects are in a small image , like 32*32 image with 10*10 object, can we reduce the stripe size to lets say 2 or 4 or 6 or 8 or 10

**Comment by djdll** [comment]
@glenn-jocher  3. Use mosaic and copy-paste augmentations during training.In you mentioned three involves the application of Mosaic, I want to let the increase in the number of _mosaic9 now supports _mosaic16, do I only need to modify ultralytics/data/augment py in the class of Mosaic can, could you please tell me where else still needs to be modified.  Thanks

**Comment by glenn-jocher** [comment]
@arun-gautham For 32x32 images with 10x10 objects, I'd recommend increasing your input size to at least 320-640 pixels using `imgsz=320` or higher. This gives the model more pixels to work with for feature extraction. You might also want to reduce the confidence threshold and experiment with different model sizes.

@djdll For adding mosaic16 support, you'd need to modify the `Mosaic` class in `ultralytics/data/augment.py`. Currently the class supports `n` values of 4 or 9 (as shown in the assertion `assert n in {4, 9}`), so you'd need to update this validation and add a `_mosaic16` method following the pattern of the existing `_mosaic4` and `_mosaic9` methods. The grid arrangement logic would need to handle a 4x4 layout for 16 images.

**Comment by djdll** [comment]
@glenn-jocher I found a suspected bug，In ultralytics/data/augment.py ->def _cat_labels(self, mosaic_labels):->imgsz = self.imgsz * 2  # mosaic imgsz ,Here either mosaic=9 or 4 is self.imgsz * 2. I think mosaic=9 should be self.imgsz * 3, otherwise some targets will be cropped. I am not sure if I am right, I hope to get a reply
